sklearn/base.py

```diff
@@ -237,6 +237,36 @@ def set_params(self, **params):
 
         return self
 
+    def fit(self, X, y=None):
+        """Fit the model from data in X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like, shape (n_samples, 1)
+            Label vector, where n_samples is the number of samples.
+
+        Notes
+        -----
+        There is a very good chance that this method
+        is overridden by mixins. In the event that occurs,
+        this base class will attempt to use the method of
+        its inherited mixin.
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+
+        if hasattr(super(), 'fit'):
+            return super().fit(X, y)
```

Comment:
We'll likely start going from `MyEst(BaseEstimator, Mixin1, Mixin2)` to `MyEst(Mixin1, Mixin2, BaseEstimator)`

I doubt that will work?



>Since this might come up, I had tried just putting in an abstract method ... However it failed hundreds of unit tests

would you have the log of the failed tests?

Reply:
Hi NicolasHug, I understand your concern. Any changes to the object hierarchy should take into account Method Resolution Order (MRO). This particular implementation should be agnostic to MRO, but I will add unit tests to prove that :). If those tests fail, then we'll have tests and I can work out a better implementation; win-win.

I do have the full log of the failed test suite, how should I share it?
Here is the test summary:

```
cmd: pytest sklearn/tests/test_common.py

platform win32 -- Python 3.6.8, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: C:\Users\lucas\dev\scikit-learn, inifile: setup.cfg
collected 5672 items
=========================== short test summary info ===========================
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator ColumnTransformer which requires parameters ['transformers']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator FeatureUnion which requires parameters ['transformer_list']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator GridSearchCV which requires parameters ['estimator', 'param_grid']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator Pipeline which requires parameters ['steps']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator RandomizedSearchCV which requires parameters ['estimator', 'param_distributions']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator SparseCoder which requires parameters ['dictionary']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator VotingClassifier which requires parameters ['estimators']
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Can't instantiate estimator VotingRegressor which requires parameters ['estimators']
SKIPPED [50] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: pandas is not installed: not testing for input of type pandas.Series to class weight.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: score_samples of BernoulliRBM is not invariant when applied to a subset.
SKIPPED [3] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Skipping check_estimators_data_not_an_array for cross decomposition module as estimators are not deterministic.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: predict of DummyClassifier is not invariant when applied to a subset.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: transform of MiniBatchSparsePCA is not invariant when applied to a subset.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: Not testing NuSVC class weight as it is ignored.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: decision_function of NuSVC is not invariant when applied to a subset.
SKIPPED [1] c:\users\lucas\appdata\local\miniconda3\envs\sklearn-env\lib\site-packages\_pytest\nose.py:25: transform of SparsePCA is not invariant when applied to a subset.
====== 242 failed, 5363 passed, 67 skipped, 56 warnings in 71.26 seconds ======
```

Those failures occurred with the following implementation of `fit()` (or therelackof :P):

```python
# up at the top
from abc import abstractmethod

# and the impl...
@abstractmethod
def fit(self, X, y=None):
    """Fit the model from data in X.

    Parameters
    ----------
    X : array-like, shape (n_samples, n_features)
        Training vector, where n_samples is the number of samples
        and n_features is the number of features.

    y : array-like, shape (n_samples, 1)
        Label vector, where n_samples is the number of samples.

    Notes
    -----
    There is a very good chance that this method
    is overridden by mixins. In the event that occurs,
    this base class will attempt to use the method of
    its inherited mixin.

    Returns
    -------
    self : object
        Returns the instance itself.
    """
    pass
```

Reply:
With the @abstractmethod approach (which honestly I'd prefer if possible) the predominant issue with the unit tests that I have run so far is (perhaps as expected) that many classes do not implement fit() at all, yet inherit from `BaseEstimator`.

Example:

`TypeError: Can't instantiate abstract class NearestNeighbors with abstract methods fit`

[`NeighborsBase`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/base.py#L106), which (at least in the `base.py` where it exists) has [three mixins](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/base.py#L859) that break up its end behavior into Unsupervised, Supervised Integer and Supervised Float, which *do* implement `fit()`. None of the Mixins explicitly implement BaseEstimator, which makes sense assuming they are used w/ `NeighborsBase`.
(As a side note, one way to resolve this specific issue would be to have the NeighborsBase simply implement a `fit()` method that does nothing but return itself, but I believe its safe to assume there was a specific reason that the existing code is the way it is).